Source/SlimDetours/Thread.c

```diff
@@ -6,107 +6,123 @@
     _Out_ PULONG SuspendedHandleCount)
 {
     NTSTATUS Status;
-    ULONG i, ThreadCount, SuspendedCount;
-    PSYSTEM_PROCESS_INFORMATION pSPI, pCurrentSPI;
-    PSYSTEM_THREAD_INFORMATION pSTI;
-    PHANDLE Buffer;
-    HANDLE ThreadHandle, CurrentPID, CurrentTID;
-    OBJECT_ATTRIBUTES ObjectAttributes = RTL_CONSTANT_OBJECT_ATTRIBUTES(NULL, 0);
-
-    /* Get system process and thread information */
-    i = _1MB;
-_Try_alloc:
-    pSPI = (PSYSTEM_PROCESS_INFORMATION)detour_memory_alloc(i);
-    if (pSPI == NULL)
-    {
-        return STATUS_NO_MEMORY;
-    }
-    Status = NtQuerySystemInformation(SystemProcessInformation, pSPI, i, &i);
-    if (!NT_SUCCESS(Status))
+    PHANDLE Buffer = NULL;
+    ULONG BufferCapacity = 0;
+    ULONG SuspendedCount = 0;
+    HANDLE CurrentTID = (HANDLE)(ULONG_PTR)NtCurrentThreadId();
+    BOOL ClosePrevThread = FALSE;
+    HANDLE ThreadHandle = NULL;
+    while (TRUE)
     {
-        detour_memory_free(pSPI);
-        if (Status == STATUS_INFO_LENGTH_MISMATCH)
+        HANDLE hNextThread;
+        Status = NtGetNextThread(NtCurrentProcess(), ThreadHandle, THREAD_ACCESS, 0, 0, &hNextThread);
+        if (ClosePrevThread)
         {
-            goto _Try_alloc;
+            NtClose(ThreadHandle);
         }
-        return Status;
-    }
 
-    /* Find current process and threads */
-    CurrentPID = (HANDLE)(ULONG_PTR)NtCurrentProcessId();
-    pCurrentSPI = pSPI;
-    while (pCurrentSPI->UniqueProcessId != CurrentPID)
-    {
-        if (pCurrentSPI->NextEntryOffset == 0)
+        if (!NT_SUCCESS(Status))
         {
-            Status = STATUS_NOT_FOUND;
-            goto _Exit;
+            if (Status == STATUS_NO_MORE_ENTRIES)
+            {
+                Status = STATUS_SUCCESS;
+            }
+            break;
         }
-        pCurrentSPI = (PSYSTEM_PROCESS_INFORMATION)Add2Ptr(pCurrentSPI, pCurrentSPI->NextEntryOffset);
-    }
-    pSTI = (PSYSTEM_THREAD_INFORMATION)Add2Ptr(pCurrentSPI, sizeof(*pCurrentSPI));
 
-    /* Skip if no other threads */
-    ThreadCount = pCurrentSPI->NumberOfThreads - 1;
-    if (ThreadCount == 0)
-    {
-        *SuspendedHandles = NULL;
-        *SuspendedHandleCount = 0;
-        Status = STATUS_SUCCESS;
-        goto _Exit;
-    }
+        ThreadHandle = hNextThread;
+        ClosePrevThread = TRUE;
+
+        THREAD_BASIC_INFORMATION BasicInformation;
+        if (!NT_SUCCESS(NtQueryInformationThread(
+            ThreadHandle,
+            ThreadBasicInformation,
+            &BasicInformation,
+            sizeof(BasicInformation),
+            NULL
+        )))
+        {
+            continue;
+        }
 
-    /* Create handle array */
-    Buffer = (PHANDLE)detour_memory_alloc(ThreadCount * sizeof(HANDLE));
-    if (Buffer == NULL)
-    {
-        Status = STATUS_NO_MEMORY;
-        goto _Exit;
-    }
+        /* Skip the current thread */
+        if (BasicInformation.ClientId.UniqueThread == CurrentTID)
+        {
+            continue;
+        }

```

Comment:
How about adding a `BOOL CurrentThreadSkipped` to record current thread is skipped or not? Thus, all `NtQueryInformationThread` calls can be skipped after current thread is found.